algopy_testing/src/algopy/primitives/bytes.py

```diff
@@ -0,0 +1,167 @@
+from __future__ import annotations
+
+import base64
+from collections.abc import Sequence
+from typing import TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from collections.abc import Iterator
+
+from algopy._constants import MAX_BYTES_SIZE
+from algopy.primitives.uint64 import UInt64
+
+# TypeError, ValueError are used for operations that are compile time errors
+# ArithmeticError and subclasses are used for operations that would fail during AVM execution
+
+
+class Bytes:
+    """
+    A python implementation of an AVM []byte
+    """
+
+    value: bytes  # underlying bytes value representing the []byte
+
+    def __init__(self, value: bytes = b"") -> None:
+        self.value = _as_bytes(value)
+
+    # @classmethod
+    # def from_int(cls, value: int = 0) -> Bytes:
+    #     return cls(value.to_bytes(8))
+
+    # @classmethod
+    # def from_str(cls, value: str) -> Bytes:
+    #     return cls(str.encode(value))
+
+    def __repr__(self) -> str:
+        return repr(bytes)
+
+    def __str__(self) -> str:
+        return str(self.value)
+
+    def __bool__(self) -> bool:
+        return bool(self.value)
+
+    def __add__(self, other: Bytes | bytes) -> Bytes:
+        """Concatenate Bytes with another Bytes or bytes literal
+        e.g. `Bytes(b"Hello ") + b"World"`."""
+        return (
+            _checked_result(self.value + other.value, "+")
+            if isinstance(other, Bytes)
+            else _checked_result(self.value + other, "+")
+        )
+
+    def __radd__(self, other: bytes) -> Bytes:
+        """Concatenate Bytes with another Bytes or bytes literal
+        e.g. `b"Hello " + Bytes(b"World")`."""
+        return _checked_result(other + self.value, "+")
+
+    def __len__(self) -> int:
+        return len(self.value)
+
+    def __iter__(self) -> Iterator[Bytes]:
+        """Bytes can be iterated, yielding each consecutive byte"""
+        return _BytesIter(self, 1)
+
+    def __reversed__(self) -> Iterator[Bytes]:
+        """Bytes can be iterated in reverse, yield each preceding byte starting at the end"""
+        return _BytesIter(self, -1)
+
+    def __getitem__(
+        self, index: UInt64 | int | slice
+    ) -> Bytes:  # maps to substring/substring3 if slice, extract/extract3 otherwise?
+        """Returns a Bytes containing a single byte if indexed with UInt64 or int
+        otherwise the substring o bytes described by the slice"""
+        # my_bytes[0:1] => b'j' whereas my_bytes[0] => 106
+        return (
+            Bytes(self.value[index])
+            if isinstance(index, slice)
+            else Bytes(self.value[slice(index, index + 1)])
+        )
+
+    def __eq__(self, other: object) -> bool:
+        maybe_bytes = _as_maybe_bytes(other)
+        # returning NotImplemented here will allow BigUInt to handle larger comparisons
```

Comment:
This comment doesn't apply for Bytes since we don't do any up-casting, instead should use `_as_bytes(other)` as it would be a compile error to compare to other types